#derive (Pread) requires From<u8> for arrays?

Perhaps I'm not understanding some elements of scroll_derive and using it wrong, but the following won't compile:
```rust
use scroll::{Pread, Pwrite};

#[derive(Pread, Pwrite)]
struct A {
    b_array: [B; 5]
}

#[derive(Pread, Pwrite, Clone, Copy)]
struct B {
    field:u32
}
```

the error is:
```
error[E0277]: the trait bound `B: From<u8>` is not satisfied
 --> src/main.rs:4:10
  |
4 | #[derive(Pread, Pwrite)]
  |          ^^^^^ the trait `From<u8>` is not implemented for `B`, which is required by `u8: Into<_>`
  |
  = note: required for `u8` to implement `Into<B>`
  = note: this error originates in the derive macro `Pread` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Adding a `From<u8>` impl on `B` does allow it to compile, but that seems like maybe not the right thing to do, since there isn't an actual real conversion from u8?
```rust
impl From<u8> for B {
    fn from(_value: u8) -> Self {
        Self { field: 0}
    }
}
```

Is there something I should do differently here to make this work as expected?

cread with also broken in same manner: https://github.com/m4b/scroll/blob/f71127293d6c90c8a1ae23c4e5981f2f48483867/scroll_derive/src/lib.rs#L422

unfortunately, something like:

```diff
diff --git a/scroll_derive/src/lib.rs b/scroll_derive/src/lib.rs
index a5714dd..846e356 100644
--- a/scroll_derive/src/lib.rs
+++ b/scroll_derive/src/lib.rs
@@ -14,18 +14,6 @@ fn impl_field(
     let default_ctx = syn::Ident::new("ctx", proc_macro2::Span::call_site()).into_token_stream();
     let ctx = custom_ctx.unwrap_or(&default_ctx);
     match *ty {
-        syn::Type::Array(ref array) => match array.len {
-            syn::Expr::Lit(syn::ExprLit {
-                lit: syn::Lit::Int(ref int),
-                ..
-            }) => {
-                let size = int.base10_parse::<usize>().unwrap();
-                quote! {
-                    #ident: { let mut __tmp: #ty = [0u8.into(); #size]; src.gread_inout_with(offset, &mut __tmp, #ctx)?; __tmp }
-                }
-            }
-            _ => panic!("Pread derive with bad array constexpr"),
-        },
         syn::Type::Group(ref group) => impl_field(ident, &group.elem, custom_ctx),
         _ => {
             quote! {
```

while it fixes the issue for non-generic structs, it unfortunately is broken for generics, seems like the error type diverges?

```
error[E0271]: type mismatch resolving `<T as TryFromCtx<'_, Endian>>::Error == Error`
   --> tests/tests.rs:177:32
    |
177 | #[derive(Debug, PartialEq, Eq, Pread, Pwrite, IOread, IOwrite, SizeWith)]
    |                                ^^^^^ expected `Error`, found associated type
    |
    = note:         expected enum `scroll::Error`
            found associated type `<T as TryFromCtx<'_, Endian>>::Error`
    = note: required for `[T; 3]` to implement `TryFromCtx<'_, Endian>`
note: required by a bound in `gread_with`
   --> /home/m4b/projects/scroll/src/pread.rs:113:26
    |
113 |     fn gread_with<'a, N: TryFromCtx<'a, Ctx, Self, Error = E>>(
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Pread::gread_with`
    = note: this error originates in the derive macro `Pread` (in Nightly builds, run with -Z macro-backtrace for more info)

```

i think there is probably an easy fix, but it's not hitting me at the moment and i didn't add the generics part so somewhat unsure about those code paths :)

I have a patch fixing this but:
```rust
impl<'a, T, Y> ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian> for Data8<T, Y>
where
    T: ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian, Error = ::scroll::Error>
        + ::std::marker::Copy,

```

I don't like how the T error is fixed to `::scroll::Error`, I feel like i should be able to remove that and do:

```rust
            ids: src
                .gread_with::<[T; 3]>(offset, ctx)
                .map_err(::scroll::Error::from)?,
```

on all the greads but it doesn't like this., with this (to my mind, very odd) error message:

```
error[E0271]: type mismatch resolving `<T as TryFromCtx<'_, Endian>>::Error == Error`
   --> tests/tests.rs:208:47
    |
208 |                 .gread_with::<[T; 3]>(offset, ctx)
    |                  ----------                   ^^^ expected `Error`, found associated type
    |                  |
    |                  required by a bound introduced by this call
    |
    = note:         expected enum `scroll::Error`
            found associated type `<T as TryFromCtx<'_, Endian>>::Error`
    = note: required for `[T; 3]` to implement `TryFromCtx<'_, Endian>`
```

unsure why it thinks `ctx` all of sudden *should* be an Error, but is instead an associated type? :thinking:

--- a/scroll_derive/src/lib.rs
+++ b/scroll_derive/src/lib.rs
@@ -14,18 +14,6 @@ fn impl_field(
     let default_ctx = syn::Ident::new("ctx", proc_macro2::Span::call_site()).into_token_stream();
     let ctx = custom_ctx.unwrap_or(&default_ctx);
     match *ty {
-        syn::Type::Array(ref array) => match array.len {
-            syn::Expr::Lit(syn::ExprLit {
-                lit: syn::Lit::Int(ref int),
-                ..
-            }) => {
-                let size = int.base10_parse::<usize>().unwrap();
-                quote! {
-                    #ident: { let mut __tmp: #ty = [0u8.into(); #size]; src.gread_inout_with(offset, &mut __tmp, #ctx)?; __tmp }
-                }
-            }
-            _ => panic!("Pread derive with bad array constexpr"),
-        },
         syn::Type::Group(ref group) => impl_field(ident, &group.elem, custom_ctx),
         _ => {
             quote! {
@@ -114,7 +102,7 @@ fn impl_struct(
             syn::GenericParam::Type(ref t) => {
                 let ident = &t.ident;
                 quote! {
-                    #ident : ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian> + ::std::convert::From<u8> + ::std::marker::Copy,
+                    #ident : ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian, Error = ::scroll::Error> + ::std::marker::Copy,
                     ::scroll::Error : ::std::convert::From<< #ident as ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian>>::Error>,
                     < #ident as ::scroll::ctx::TryFromCtx<'a, ::scroll::Endian>>::Error : ::std::convert::From<scroll::Error>
                 }

--- a/scroll_derive/tests/tests.rs
+++ b/scroll_derive/tests/tests.rs
@@ -180,7 +180,6 @@ struct Data8<T, Y> {
     ids: [T; 3],
     xyz: Y,
 }
-
 #[test]
 fn test_generics() {
     let mut bytes = [0xde, 0xed, 0xef, 0x10, 0x10];
@@ -268,4 +267,18 @@ fn test_custom_ctx_derive() {
         c: 0x0506,
     };
     assert_eq!(data3, bytes.ioread_with(LE).unwrap());
+}
+
+#[derive(Debug, Pread, SizeWith)]
+struct Data12 {
+    ids: [Data11; 1],
+}
+
+#[test]
+fn test_array_with_nested_pread_data() {
+    let bytes = [0xde, 0xed, 0xef, 0x10, 0x10, 0x01];
+    let data: Data12 = bytes.pread_with(0, LE).unwrap();
+    assert_eq!(data.ids[0].a, 0xedde);
+    assert_eq!(data.ids[0].b, 0x10ef);
+    assert_eq!(data.ids[0].c, 0x1001);
 }

